src/metrics.py

```diff
@@ -0,0 +1,301 @@
+# Copyright 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+#  * Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+#  * Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#  * Neither the name of NVIDIA CORPORATION nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from typing import Dict, Union
+
+import triton_python_backend_utils as pb_utils
+from vllm.engine.metrics import StatLoggerBase as VllmStatLoggerBase
+from vllm.engine.metrics import Stats as VllmStats
+from vllm.engine.metrics import SupportsMetricsInfo
+
+
+# begin-metrics-definitions
+class TritonMetrics:
+    def __init__(self, labels):
```

Comment:
Do we need to verify if server metrics are enabled? I think if you build server with `TRITON_ENABLE_METRICS=OFF`, then the vLLM backend will become unusable unless you put some safeguards to only use metrics if enabled.

Reply:
@rmccorm4 Fix for (a).
- https://github.com/triton-inference-server/vllm_backend/pull/53/commits/979dc02bc979362f7f28a39b0cd75320fb905b88
- https://github.com/triton-inference-server/python_backend/pull/375

<img width="1491" alt="Screenshot 2024-08-07 at 3 19 20 PM" src="https://github.com/user-attachments/assets/99969dcb-6f37-4d22-a286-bc32156b3a47">

Reply:
(b) I ran some inference and it looks normal to me. No crash no exception.

Reply:
Metrics are still created in (b). Move the fix to the optimization ticket.